### DISCUSSION TITLE
Slow Inference Speeds on Cluster GPUs

### DISCUSSION THREAD

**Discussion by jmdelahanty** [talmolab/sleap#833] [Help!] [answered]
Hello everyone!

Something that has seemed to happen in our lab's use of SLEAP while on the SNL cluster is that our inference speeds are around 30FPS or so on a GPU.

From the [methods paper](https://www.nature.com/articles/s41592-022-01426-1) (which I think had things trained on A100 GPUs?) there's reported speeds of several hundred FPS! Crazy cool speed!

While we aren't using A100s (yet), I would think that 30FPS is somewhat slow. Here's the specs of an example machine people are running on:

```
(base) jdelahanty@tesla:/snlkt/data/facial_expression/specialk$ nvidia-smi -l 1
Fri Jul  8 13:41:15 2022
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.106.00   Driver Version: 460.106.00   CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla P100-PCIE...  Off  | 00000000:04:00.0 Off |                    0 |
| N/A   33C    P0    34W / 250W |   8353MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P100-PCIE...  Off  | 00000000:05:00.0 Off |                    0 |
| N/A   34C    P0    34W / 250W |   8353MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  Tesla P100-PCIE...  Off  | 00000000:08:00.0 Off |                    0 |
| N/A   30C    P0    32W / 250W |   8353MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  Tesla P100-PCIE...  Off  | 00000000:09:00.0 Off |                    0 |
| N/A   31C    P0    33W / 250W |   8353MiB / 16280MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   4  Tesla K40c          Off  | 00000000:88:00.0 Off |                    0 |
| 23%   26C    P8    20W / 235W |      3MiB / 11441MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A     37797      C   ...da3/envs/sleap/bin/python     8351MiB |
|    1   N/A  N/A     38257      C   ...da3/envs/sleap/bin/python     8351MiB |
|    2   N/A  N/A      1241      C   ...da3/envs/sleap/bin/python     8351MiB |
|    3   N/A  N/A      6198      C   ...da3/envs/sleap/bin/python     8351MiB |
```

You can even see that there's some people running sleap on it!

Here's some info about the machine itself:

```
tesla is managed by SNL Support <support@salk.edu> and runs Debian Stretch v9.X.
CPU Count: 40 : "GenuineIntel Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (2 chips x 10 cores : 20 hyperthread cores)"
```

Since this computer only has 20 cores, so 40 with hyperthreading, maybe that could be a bottleneck for it? It also looks like the batch size of the inferences could be almost doubled since since lots of memory is available.

Perhaps the most important thing to notice is that the `GPU-Util` value is 0%! I have a feeling that even though people are trying to specify GPU usage, the GPU isn't really doing much work! I'm basing my understanding of what the `GPU-Util` parameter says from [this Medium post](https://medium.com/analytics-vidhya/explained-output-of-nvidia-smi-utility-fc4fbee3b124) and [these docs](https://nvidia.custhelp.com/app/answers/detail/a_id/3751/~/useful-nvidia-smi-queries).

Here's an attempt to query things and it's output:

```
(base) jdelahanty@tesla:/snlkt/data/facial_expression/specialk$ nvidia-smi --query-gpu=timestamp,pci.bus_id,utilization.gpu,driver_version,name --format
=csv
timestamp, pci.bus_id, utilization.gpu [%], driver_version, name
2022/07/08 14:14:25.791, 00000000:04:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/07/08 14:14:25.796, 00000000:05:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/07/08 14:14:25.801, 00000000:08:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/07/08 14:14:25.805, 00000000:09:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/07/08 14:14:25.807, 00000000:88:00.0, 0 %, 460.106.00, Tesla K40c
```

Edit:

After staring at nvidia-smi polling for a few minutes, sometimes the GPU-Util gets to 15% for a second, then drops down to zero again.

Edit2:

It could be that they were running an older version of SLEAP that had latency between the CPU and GPU from a bug that Arlo mentioned he fixed. It looks like on most recent versions of SLEAP the GPU is getting utilized, but only gets up to something like 50%...

I could be misinterpreting things, but it looks to me that things are being loaded into the GPU but the card isn't actually doing anything for some reason!

The `rate` they're seeing printed to the terminal shows a value hovering around 5. I'm not sure what that rate means. Is it literally the frame rate inference?

I was asking @sheridana about this briefly today and he suggested a couple things to check:

- [ ] Try doing inference when the data is stored on the local machine instead of being read over the network
- [ ] Try just running the video with all default settings and see if it's something particular about the settings we're using
- [ ] See if it depends on how much CPU is available, its possible that there's a lot of latency that builds up over time when transferring data from CPU to GPU
- [ ] Look into whether or not the GPU is actually even being used (which it looks like it might not be!)

Any advice?

**Comment by talmo** [comment]
The most likely bottleneck is reading the data, especially if it's stored over the network.

The CPU stuff is very unlikely -- most of the heavy lifting in SLEAP is done on the GPU, so 20 cores is way overkill.

Other things to try:
- Don't enable the tracker to see if the performance is limited by the model inference alone.
- Increase the batch size. P100s are pretty big in terms of memory, so try 16 or 32.
- Try training a smaller model. Model sizes of ~1-4 million parameters are considered very lightweight and should get you triple digit FPS.
- If possible, downscale your images by 0.5 or 0.75x if you won't lose resolution.
- Depending on your use case, implementing a [custom inference pipeline](https://sleap.ai/notebooks/Interactive_and_realtime_inference.html) can optimize things if you want to handle the data loading yourself in a more specialized way.

**Reply by jmdelahanty** [reply]
The first thing I'll do is have people try making a copy of the data on the compute's local drive.

I find that confusing though since I thought the connections there are 10-20Gbps and so streaming the data over the network would be sufficiently fast for getting things efficiently loaded into the machine. Am I misunderstanding how networked systems like this can be used? How does VAST get around this problem of moving data around?

**Reply by jmdelahanty** [reply]
Even when reading from the local machine's scratch space for the video (which I'm told has an SSD on board!) I'm still only getting 4% GPU utilization! Will post here about the command that was used for `sleap-track`...

**Reply by talmo** [reply]
Hey Jeremy,

Can you try preloading the video frames to exclude I/O when benchmarking? Something like:

```py
import sleap
import numpy as np
from time import perf_counter

video = sleap.load_video("video.mp4")
predictor = sleap.load_model("path/to/model_folder", batch_size=16)

# Preload images
imgs = video[:1024]

predictor.predict(imgs[:16])  # warmup

for _ in range(3):  # do multiple trials
    t0 = perf_counter()
    predictor.predict(imgs)
    elapsed = perf_counter() - t0

    fps = len(imgs) / elapsed
    print(fps)
```

This will give you a baseline for how much time is being spent on the SLEAP side doing inference. If there's a big discrepancy with what you're seeing, it might be either the tracker outside of the neural network or your videos are slow to decode for some reason.

**Reply by jmdelahanty** [reply]
Getting stuck somehow here with this traceback:

```
 predictor.predict(imgs[:16])
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% ETA: -:--:-- ?
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 436, in predict
    self._make_labeled_frames_from_generator(generator, data)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 2126, in _make_labeled_frames_from_generator
    for ex in generator:
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 346, in _predict_generator
    ex = process_batch(ex)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 312, in process_batch
    preds = self.inference_model.predict_on_batch(ex, numpy=True)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 916, in predict_on_batch
    outs = super().predict_on_batch(data, **kwargs)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/keras/engine/training.py", line 1947, in predict_on_batch
    outputs = self.predict_function(iterator)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 885, in __call__
    result = self._call(*args, **kwds)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 924, in _call
    results = self._stateful_fn(*args, **kwds)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3038, in __call__
    filtered_flat_args) = self._maybe_define_function(args, kwargs)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3460, in _maybe_define_function
    args, kwargs, flat_args, filtered_flat_args, cache_key_context)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3382, in _define_function_with_shape_relaxation
    args, kwargs, override_flat_arg_shapes=relaxed_arg_shapes)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3308, in _create_graph_function
    capture_by_value=self._capture_by_value),
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/framework/func_graph.py", line 1007, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 668, in wrapped_fn
    out = weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/framework/func_graph.py", line 994, in wrapper
    raise e.ag_error_metadata.to_exception(e)
tensorflow.python.autograph.impl.api.StagingError: in user code:

    /home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/keras/engine/training.py:1586 predict_function  *
        return step_function(self, iterator)
    /home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py:1848 call  *
        peaks_output = self.instance_peaks(example, crop_output)
    /home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py:692 call  *
        a = tf.expand_dims(

    KeyError: 'instances'
```

Any tips?

**Reply by jmdelahanty** [reply]
For posterity also, here's an example shell script command that a user in the lab is running where the same behavior occurs with slow prediction/tracking speed:

```
sleap-track /nadata/snlkt/spb_psilicon/Pilot_BehParameters/Videos_RawData/FFBatch/20220619_Con1_3934R_M_RL.mp4 -m /nadata/snlkt/spb_psilicon/Pilot_BehParameters/SLEAP/SLEAP_phase4/models/220801_174806.centroid.n=2092/training_config.json -m /nadata/snlkt/spb_psilicon/Pilot_BehParameters/SLEAP/SLEAP_phase4/models/220801_181533.centered_instance.n=2092/training_config.json --gpu 0  --tracking.tracker flow --tracking.pre_cull_to_target 0 --tracking.pre_cull_iou_threshold 0.8 --tracking.similarity instance --tracking.match greedy --tracking.track_window 30 --tracking.post_connect_single_breaks 0 -o /nadata/snlkt/spb_psilicon/Pilot_BehParameters/BehAnalysis/SLEAP_processed_Pilot_BehParameters_v2/predictions/20220619_Con1_3934R_M_RL.predictions.slp --verbosity json --no-empty-frames; sleap-convert --format analysis -o "/nadata/snlkt/spb_psilicon/Pilot_BehParameters/BehAnalysis/SLEAP_processed_Pilot_BehParameters_v2/predictions/20220619_Con1_3934R_M_RL.predictions.analysis.h5" "/nadata/snlkt/spb_psilicon/Pilot_BehParameters/BehAnalysis/SLEAP_processed_Pilot_BehParameters_v2/predictions/20220619_Con1_3934R_M_RL.predictions.slp";
```

**Reply by talmo** [reply]
> Getting stuck somehow here with this traceback:
> 
> ```
>  predictor.predict(imgs[:16])
> Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% ETA: -:--:-- ?
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 436, in predict
>     self._make_labeled_frames_from_generator(generator, data)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 2126, in _make_labeled_frames_from_generator
>     for ex in generator:
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 346, in _predict_generator
>     ex = process_batch(ex)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 312, in process_batch
>     preds = self.inference_model.predict_on_batch(ex, numpy=True)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py", line 916, in predict_on_batch
>     outs = super().predict_on_batch(data, **kwargs)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/keras/engine/training.py", line 1947, in predict_on_batch
>     outputs = self.predict_function(iterator)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 885, in __call__
>     result = self._call(*args, **kwds)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 924, in _call
>     results = self._stateful_fn(*args, **kwds)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3038, in __call__
>     filtered_flat_args) = self._maybe_define_function(args, kwargs)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3460, in _maybe_define_function
>     args, kwargs, flat_args, filtered_flat_args, cache_key_context)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3382, in _define_function_with_shape_relaxation
>     args, kwargs, override_flat_arg_shapes=relaxed_arg_shapes)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 3308, in _create_graph_function
>     capture_by_value=self._capture_by_value),
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/framework/func_graph.py", line 1007, in func_graph_from_py_func
>     func_outputs = python_func(*func_args, **func_kwargs)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 668, in wrapped_fn
>     out = weak_wrapped_fn().__wrapped__(*args, **kwds)
>   File "/home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/tensorflow/python/framework/func_graph.py", line 994, in wrapper
>     raise e.ag_error_metadata.to_exception(e)
> tensorflow.python.autograph.impl.api.StagingError: in user code:
> 
>     /home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/keras/engine/training.py:1586 predict_function  *
>         return step_function(self, iterator)
>     /home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py:1848 call  *
>         peaks_output = self.instance_peaks(example, crop_output)
>     /home/jdelahanty/miniconda3/envs/sleap/lib/python3.7/site-packages/sleap/nn/inference.py:692 call  *
>         a = tf.expand_dims(
> 
>     KeyError: 'instances'
> ```
> 
> Any tips?

You have to load both models when doing top-down:

```py
predictor = sleap.load_model([
    "/nadata/snlkt/spb_psilicon/Pilot_BehParameters/SLEAP/SLEAP_phase4/models/220801_174806.centroid.n=2092",
    "/nadata/snlkt/spb_psilicon/Pilot_BehParameters/SLEAP/SLEAP_phase4/models/220801_181533.centered_instance.n=2092",
], batch_size=16)
```

Admittedly, we could probably have a more informative error message :)

**Reply by jmdelahanty** [reply]
That worked! Didn't know about needing both for top-down. Here's what I got:

```
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 90.0 FPS
Labels(labeled_frames=1024, videos=1, skeletons=1, tracks=0)
78.12063462791669
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 125.7 FPS
Labels(labeled_frames=1024, videos=1, skeletons=1, tracks=0)
118.06317465381672
Predicting... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% ETA: 0:00:00 130.9 FPS
Labels(labeled_frames=1024, videos=1, skeletons=1, tracks=0)
120.68165525608238
```

Much better frame rate and we could definitely up the batch size I think and help it go even faster! So how can I go about testing if it's the file I/O or the tracker outside the network slowing things down?

Also have no idea if this is relevant, but here's what I get when I just ask what the `video` is:

```
Video(backend=MediaVideo(filename='/scratch/spb_benchmark/20220628_Con4_3934R_M_RL.mp4', grayscale=False, bgr=True, dataset='', input_format='channels_last'))
```

~I was surprised that the backend says `MediaVideo` and not something like `ffmpeg` but I also have no idea what backend really refers to...~

MediaVideo uses opencv which relies on ffmpeg in the backend! Talmo taught me that a few minutes ago.

**Reply by talmo** [reply]
You can try doing each separately.

Time how long the video loading takes when doing it batch by batch:
```py
video = sleap.load_video("video.mp4")
batch_size = 16
n_frames = 4096 # or len(video), but this might take a while
for i0 in range(0, n_frames, batch_size):
    t0 = perf_counter()
    i1 = min(i0 + batch_size, len(video))
    imgs = video[i0:i1]
    elapsed = perf_counter() - t0
    fps = len(imgs) / elapsed
    print(fps)
```

And the tracking using the approach in [this notebook](https://sleap.ai/notebooks/Post_inference_tracking.html).

**Reply by jmdelahanty** [reply]
Quick update! It looks like the loading piece is quite fast both from a remote source as well as the local source! Here's what that looked like:

```
# Some users are experiencing slow video IO over our cluster network (probably)
# This script just loads first a local scratch space video and times how long it takes
# to read through each batch of frames. It will then do it from remote
# storage to see how long it takes to access the frames.

import sleap
from time import perf_counter
import numpy as np


video = sleap.load_video("/scratch/spb_benchmark/20220628_Con4_3934R_M_RL.mp4")
batch_size = 16
n_frames = 4096 # or len(video), but this might take a while
local_fps = []
for i0 in range(0, n_frames, batch_size):
    t0 = perf_counter()
    i1 = min(i0 + batch_size, len(video))
    imgs = video[i0:i1]
    elapsed = perf_counter() - t0
    fps = len(imgs) / elapsed
    print("LOCAL: ", fps)
    local_fps.append(fps)

remote = sleap.load_video("/snlkt/spb_psilicon/Pilot_BehParameters/Videos_RawData/FFBatch/20220619_Con1_3934R_M_RL.mp4")
remote_fps = []
for i0 in range(0, n_frames, batch_size):
    t0 = perf_counter()
    i1 = min(i0 + batch_size, len(video))
    imgs = remote[i0:i1]
    elapsed = perf_counter() - t0
    fps = len(imgs) / elapsed
    print("REMOTE: ", fps)
    remote_fps.append(fps)

print(np.mean(np.array(local_fps)))
print(np.mean(np.array(remote_fps)))
```

Avg FPS Local: 559
Avg FPS Remote: 569

Very cool that things are accessed that fast!

EDIT:

Just to see if maybe things started out fast but were slowing down later, I ran `predict` on both a local and remote source and things were equivalently speedy with maintained performance throughout from what I could tell!

FPS Local: 252
FPS Remote: 250

**Reply by jmdelahanty** [reply]
Potentially dumb question:

I'm not sure how to save a `predictions.slp` file from a Python script. So using the prediction script you made above, I'm not sure how to actually iterate over a video and append to a `.slp` file that I can then apply the tracking method to later. Or would it best to just have a video where predictions have been done already and do tracking on that?

Secondary potentially dumb question:

Should users run predictions first upon the whole video and then do the tracking? Or is that handled all in one CLI call? Or does `sleap-track` also perform predictions? ~Does `sleap-track` even use a GPU for things?~ <-- Yes it does, just still doesn't seem to use it that often...

**Reply by jmdelahanty** [reply]
Laurel happens to be running a bunch of sleap on Tesla right now! The GPUs are barely getting used but the CPU is popping off! Here's a screenshot:

![image](https://user-images.githubusercontent.com/28411193/183032432-65cd4d72-ccf5-4b6d-a816-125ba8558407.png)

Meanwhile the GPU polling...

```
./nvidia-monitor.sh
timestamp, pci.bus_id, utilization.gpu [%], driver_version, name
2022/08/05 01:09:00.876, 00000000:04:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:00.877, 00000000:05:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:00.877, 00000000:08:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:00.879, 00000000:09:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:00.879, 00000000:88:00.0, 0 %, 460.106.00, Tesla K40c
2022/08/05 01:09:01.880, 00000000:04:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:01.881, 00000000:05:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:01.881, 00000000:08:00.0, 0 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:01.881, 00000000:09:00.0, 15 %, 460.106.00, Tesla P100-PCIE-16GB
2022/08/05 01:09:01.882, 00000000:88:00.0, 0 %, 460.106.00, Tesla K40c
```

**Reply by jmdelahanty** [reply]
Lastly, question about this:

>  sleap-track is the command-line interface for running inference using models which have already been trained. Use this for running inference on a remote machine such as an HPC cluster or Colab notebook.
>
> If you specify how many identities there should be in a frame (i.e., the number of animals) with the --tracking.clean_instance_count argument, then we will use a heuristic method to connect “breaks” in the track identities where we lose one identity and spawn another. This can be used as part of the inference pipeline (if models are specified), as part of the tracking-only pipeline (if the predictions file is specified and no models are specified), or by itself on predictions with pre-tracked identities (if you specify --tracking.tracker none). See [Tracking and proofreading](https://sleap.ai/guides/proofreading.html#proofreading) for more details on tracking.

Does this imply that predictions should be done first on an entire video and then the tracking done afterwards? Or if both happen does the program do predictions and then track those predictions while those video frames are still in the GPU's memory? Predictions are held in `.slp` files too right? So maybe slowdowns happen by there being multiple kinds of I/O happening over a network or something?

**Reply by jmdelahanty** [reply]
~For tracking, it looks like it's dependent entirely on CPU. Is this the case or am I just using things incorrectly?~ If I looked more carefully at the notebook, I would have seen that it is a CPU bound job. 
> Note: Tracking does not run on the GPU, so this notebook can be run locally on your computer without the hassle of uploading your data if desired.

It would make sense that if people are performing both tracking *and* predicting on videos all at once that the machine just gets super busy and slows *wayyyyyy* down. From doing a test on a machine with nothing else running, doing the tracking procedure outlined in the notebook uses up almost every core/thread! People have shell scripts that are executing several videos all at the same time. I'm willing to be that's the issue probably with how people are getting these slowdowns to 5 FPS (which Laurel confirmed that she's seeing similar things today!).

Is there a way to limit the number of threads sleap is capable of using for this part of the tracking process without throwing in an environment variable every time in the shell script (like `OMP_NUM_THREADS`) for it. Which would also require the `threadpoolctl` library.

I know of [one user friendly way](https://www.run.ai/) of containerization that could do that sort of thing...

But would making a docker image and spinning up containers in parallel be the way forward until orchestration software is adopted? I have some experience with doing that for getting parallel preprocessing going from adapting a repo from the Deisseroth lab that you can see [here](https://github.com/Tyelab/bruker_pipeline/blob/main/docker/beyblade.py)

EDIT:

It looks like the tracking piece is what takes so long, not the prediction part. At least from what I'm experiencing testing at the moment... tracking with the parameters used in the notebook on a video with around 200k frames took just over 40 minutes. Here's the hardware I ran it on:

```
poseidon is managed by SNL Support <support@salk.edu> and runs Debian Buster v10.X.
CPU Count: 48 : "AuthenticAMD AMD Ryzen Threadripper 3960X 24-Core Processor 2195.958 MHz (1                                                                                                                  chips x 24 cores : 48 hyperthread cores)"
```

This makes me feel that the main reason for the slowdowns people are experiencing is that they're trying to run these heavy CPU operations without limiting any of their trackers' resource usage. Sometimes people try to several videos at once from a shell script and I think it's just overwhelming the machines probably.

Performance for all the other steps appears to be quite nice!

**Reply by roomrys** [reply]
> Does this imply that predictions should be done first on an entire video and then the tracking done afterwards? Or if both happen does the program do predictions and then track those predictions while those video frames are still in the GPU's memory? Predictions are held in .slp files too right? So maybe slowdowns happen by there being multiple kinds of I/O happening over a network or something?

The predictions are always done before tracking (at the moment - we are working on adding tracks as classifiers, but that feature will only be released later down the line). The `sleap-track` command can generate prediction and then add tracks all in one command. I am currently working on a fix for tracking pre-existing predictions through `sleap-track` (see #867). Predictions are added to the current `.slp`, but also output to a `.predictions.slp` which contains only the predictions.

**Reply by talmo** [reply]
I'll just jump in here to say that tracking in general is slow. This is fundamentally because you have to solve a ton of matching problems via the Hungarian algorithm -- basically once per frame. To make matters worse, since tracking also depends on previous frames, it's also non-trivial to parallelize it.

One thing that may help here is to use "simple" instead of "flow" tracking, though the results might be worse and the speed ups probably not too great.

If you can separate GPU vs CPU workloads, that's probably a good move for now in terms of resource management and the threads issue.

**Reply by jmdelahanty** [reply]
I'll give this a shot this week Talmo and see what happens! Also cool to know more about what's being used under the hood, I need to look into Hungarian matching algorithms (I'm Hungarian which is fun!). Also looking into how you can parallelize it would be neat, although I have no idea how to successfully do that.

It looks like doing a `sleap-track` for predicting on frames and then doing `sleap-track` for the tracking piece helps things stay pretty speedy, it's just that the tracking part is using up almost all the cores on just one instance of `sleap-track` still. Without containerization I'm not sure there's an easy way to do that efficiently...

And for @roomrys answer, is this just the `retrack` piece or is it tracking in general? I was able to run a tracker successfully through the command line as far as I can tell!

**Reply by roomrys** [reply]
@jmdelahanty, the fix that I mentioned was just for retracking on already existing predictions.

**Reply by jmdelahanty** [reply]
Cool update here! So something that appears to have helped quite a lot potentially is the use of the correct drivers on our system. Will update soon!

**Reply by jmdelahanty** [reply]
Okay, it looks like separating out the CPU and GPU workloads is helpful. On our new A40 servers, some users are getting 350+ FPS even while running 4 cards simultaneously!

Second, it appears that on our cluster that there's multiple drivers available on the system since different software relies upon different drivers depending on when/how it was built. If we tell the bash environment to reference the correct drivers things run without any issues! This also solved the problem that @sciencemassa was running into with her predictions/training I believe [see [here](https://github.com/talmolab/sleap/discussions/894) for discussion]. Interesting that things don't halt if the right driver library can't be found sometimes! So this is probably a solved problem now!

**Reply by jmdelahanty** [reply]
Turns out this did not solve the issue in the linked discussion. It's happening sporadically and only to one specific project in the lab. Super odd! However, separating out the GPU and CPU workloads has dramatically improved performance. On our cluster's A40s, we're achieving inference speeds of 400+FPS when only inference is done and the tracking portion takes just a couple minutes to complete on CPU. Things that were once taking a bunch of time (on the order of something like 4 hours!) now only takes like 10 minutes! Amazing!

**Reply by boadecea25** [reply]
> You can try doing each separately.
> 
> Time how long the video loading takes when doing it batch by batch:
> 
> ```python
> video = sleap.load_video("video.mp4")
> batch_size = 16
> n_frames = 4096 # or len(video), but this might take a while
> for i0 in range(0, n_frames, batch_size):
>     t0 = perf_counter()
>     i1 = min(i0 + batch_size, len(video))
>     imgs = video[i0:i1]
>     elapsed = perf_counter() - t0
>     fps = len(imgs) / elapsed
>     print(fps)
> ```
> 
> And the tracking using the approach in [this notebook](https://sleap.ai/notebooks/Post_inference_tracking.html).

Hello @talmo , I did this and with preloading frames, sleap's inference remains the same despite excluding tracking as well. What could be the issue in this case?

**Reply by jmdelahanty** [reply]
I think what this snippet does is just test what the fps is for just reading video frames from a video. The FPS that finds should tell you about what loading videos is like but doesn't change what the inference step is doing. This just can tell you if there's a bottleneck in loading video frames or something else I'm pretty sure.

**Reply by boadecea25** [reply]
So I ran a benchmarking code for all (loading videos, inference by sleap with a preloaded video and tracking). Surprisingly, the fps for loading video and tracking was around 250 while just sleap was 11? Is this an innate problem with sleap/ my gpu/ what could it be?
From this what I understand is the bottleneck is the inference part by itself which is unexpected.

**Reply by jmdelahanty** [reply]
The only thing I can think of that could be a problem is driver compatibility for something like that. Unfortunately for me, I'm not so experienced with debugging driver issues and I'm not even sure where to start for determining the best ones for things! Maybe @talmo has some advice?

**Comment by jmdelahanty** [comment, marked as answer]
To simplify the info in the discussion here, two things really helped us achieve speed ups from 5-10 FPS to over 400FPS:

1. Correct drivers: SLEAP/other GPU dependent software doesn't seem to crash as much as struggle if the non-optimal drivers are being referenced in your linux terminal.
2. Separating out CPU and GPU workloads as @talmo suggested. This helps enormously with people's workflows. On A40s it appears that doing both CPU/GPU tasks in the same sleap-track call is sufficiently fast, but each step is about 2x as fast if they are separated out.
